sound/soc/sof/core.c

```diff
@@ -95,6 +95,27 @@ void snd_sof_get_status(struct snd_sof_dev *sdev, u32 panic_code,
 }
 EXPORT_SYMBOL(snd_sof_get_status);
 
+/* Helper to manage DSP state */
+void sof_set_fw_state(struct snd_sof_dev *sdev, enum sof_fw_state new_state)
+{
+	if (sdev->fw_state == new_state)
+		return;
+
+	dev_dbg(sdev->dev, "fw_state change: %d -> %d\n", sdev->fw_state, new_state);
+	sdev->fw_state = new_state;
+
+	switch (new_state) {
+	case SOF_FW_BOOT_NOT_STARTED:
+	case SOF_FW_BOOT_COMPLETE:
+	case SOF_FW_CRASHED:
+		sof_client_fw_state_dispatcher(sdev);
+		fallthrough;
+	default:
+		break;
```

Comment:
can you add a comment on why the other states are discarded? Why not leave this selection to the client itself?

Reply:
I will add a comment. The reason is that the remaining `SOF_FW_BOOT_PREPARE`, `SOF_FW_BOOT_IN_PROGRESS`, `SOF_FW_BOOT_FAILED` and `SOF_FW_BOOT_READY_FAILED` is really just for the core's internal state 'tracking' and for the clients they mean nothing.
It is the same for them if the fw_state is `BOOT_NOT_STARTED`/`BOOT_PREPAR`/`BOOT_IN_PROGRESS`/`BOOT_FAILED`/`BOOT_READY_FAILED`. It all means the firmware is not booted, it is not usable.

Reply:
maybe, but if the client needs to prepare something ahead of time, they could try to branch-predict assuming the boot is fine. that'd be a way to safe initialization/resume time.

Reply:
OK, let's move the filtering to the client side.